### DISCUSSION TITLE
2022-46: OpenDAL 的错误处理实践

### DISCUSSION THREAD

**Discussion by giscus** [Xuanwo/blog#389] [Announcements]
# 2022-46: OpenDAL 的错误处理实践

An infrastructure engineer, focused on distributed storage system

https://xuanwo.io/reports/2022-46/

**Comment by sleepymole** [comment]
> Persistent 表示错误曾经是临时的，但是在重试之后还在持续出错，不鼓励用户再尝试这个请求

对 Persistent 的作用有些疑问，如果重试是由用户发起的，那 OpenDAL 是如何判断一个请求是一个重试请求的呢？另外，如果不鼓励用户重试，为什么不干脆返回 Permanent，这里期望用户针对这个 status 做什么不同的处理呢？

**Reply by Xuanwo** [reply]
> 对 Persistent 的作用有些疑问，如果重试是由用户发起的，那 OpenDAL 是如何判断一个请求是一个重试请求的呢？

好问题。

OpenDAL 实际上并不判断请求是否是重试请求，Error 的状态是由用户或者中间件来维护的。底层的 Service 总是会返回 `Permanent` 或者 `Temporary`，但是用户可以自行重试或者使用 [RetryLayer](https://opendal.databend.rs/opendal/layers/struct.RetryLayer.html) 在重试失败后将 Error 状态设置为 `Persistent` （或者 `Permanent`）。

这里额外增加了一个 `Persistent` 是为了能更清晰的标记出这个错误（的操作）已经被重试过了。

不过 `Persistent` 好像意义确实不大，后面可以考虑直接转成 `Permanent` 然后在 context 里面增加一个 already retied～

**Comment by yzhaoyu** [comment]
大佬问一下，OpenDAL对底层对象存储的访问使用的接口和外部用户访问使用OpenDAL的接口是两套接口吗？有这方面相关的设计文档吗？

**Reply by Xuanwo** [reply]
我在这里 https://opendal.databend.rs/concepts.html 简单的介绍了一下。

OpenDAL 内部使用 [Accessor](https://opendal.databend.rs/opendal/trait.Accessor.html) 来统一存储服务的抽象，但是对用户暴露的是 [Object](https://opendal.databend.rs/opendal/struct.Object.html)。OpenDAL 比较鼓励用户使用 Object 而不是直接用裸的 Accessor 接口。

**Reply by yzhaoyu** [reply]
Accessor怎么管理Object的呢？

**Reply by Xuanwo** [reply]
Accessor 是一个 rust 中的 `trait`（类似 Golang 中的 interface），不同的 Service 都实现了 `Accessor`，然后 Object 是上层的封装。以 s3 为例，`object.read()` 会通过 `Accessor` 的 `read` 接口调用 s3 服务的 `get_obejct` 接口。

**Reply by yzhaoyu** [reply]
数据的每次读写都是直接操作对象存储吗，有加一层cache做缓存吗？

**Reply by Xuanwo** [reply]
> 数据的每次读写都是直接操作对象存储吗，有加一层cache做缓存吗？

是的。OpenDAL 提供了一套中间件机制可以做这样的扩展，目前提供了几种比较简单的 cache 模式： [ContentCacheLayer](https://opendal.databend.rs/opendal/layers/struct.ContentCacheLayer.html)

**Comment by yzhaoyu** [comment]
Accessor是类似于Golang里的interface的概念吧？

**Comment by Endle** [comment]
我有一个不太成熟的想法，是不是设计一个表达错误信息的数据结构，是非常困难的事情

当程序正常运行时，程序的状态s 可以看作是某个集合 S 中的元素，集合S的内容是已知且可控的。

而当程序进入异常状态时，我们难以事先圈定一个集合 E 来涵盖所有的错误 e。

我想到之前我写的一些 Java 程序，遇到数据库操作错误时，基本都是 打一行log，再把异常丢出去，给上一层处理

**Comment by lightsing** [comment]
但是这样每次错误都会产生多个alloc，包括 HashMap 还有 Vec？是否可以选择一些更轻量的？

**Reply by Xuanwo** [reply]
可以的，这个就根据自己的业务需要来调整了🥰
